src/main/scala/org/neo4j/spark/service/SchemaService.scala

```diff
@@ -0,0 +1,140 @@
+import org.neo4j.cypherdsl.core.Cypher
+import org.neo4j.cypherdsl.core.renderer.Renderer
+import org.neo4j.driver.exceptions.ClientException
+import org.neo4j.driver.Session
+import org.neo4j.spark.util.Neo4jUtil
+import org.neo4j.spark.{DriverCache, Neo4jOptions, Neo4jQuery}
+
+import collection.JavaConverters._
+import scala.collection.mutable
+
+class SchemaService(private val options: Neo4jOptions, private val jobId: String) extends AutoCloseable {
+
+  import SchemaService._
+
+  import org.neo4j.spark.QueryType._
+
+  private val cypherRenderer = Renderer.getDefaultRenderer
+
+  private val driverCache: DriverCache = new DriverCache(options.connection, jobId)
+
+  private val session: Session = driverCache.getOrCreate().session(options.session.toNeo4jSession)
+
+  private def cypherToSparkType(cypherType: String): DataType = {
+    cypherType match {
+      case "Boolean" => DataTypes.BooleanType
+      case "String" => DataTypes.StringType
+      case "Long" => DataTypes.IntegerType
+      case "Double" => DataTypes.DoubleType
+      case "Point" | "InternalPoint2D" | "InternalPoint3D" => pointType
+      case "LocalDateTime" | "LocalTime" | "DateTime" | "ZonedDateTime" | "OffsetTime" | "Time" => timeType
+      case "LocalDate" | "Date" => DataTypes.DateType
+      case "Duration" | "InternalIsoDuration" => durationType
+      case "StringArray" => DataTypes.createArrayType(DataTypes.StringType)
+      case "DurationArray" | "InternalIsoDurationArray" => DataTypes.createArrayType(durationType)
+      case "LongArray" => DataTypes.createArrayType(DataTypes.IntegerType)
+      case "DoubleArray" => DataTypes.createArrayType(DataTypes.DoubleType)
+      case "BooleanArray" => DataTypes.createArrayType(DataTypes.BooleanType)
+      case "PointArray" | "InternalPoint2DArray" | "InternalPoint3DArray" => DataTypes.createArrayType(pointType)
+      case "LocalDateTimeArray" | "DateTimeArray" | "ZonedDateTimeArray" | "OffsetTimeArray"
+           | "TimeArray" | "LocalTimeArray" => DataTypes.createArrayType(timeType)
+      case "LocalDateArray" | "DateArray" => DataTypes.createArrayType(DataTypes.DateType)
+      case _ => DataTypes.StringType
+    }
+  }
+
+  def queryForNode(): StructType = {
+    var structFields: mutable.Buffer[StructField] = (try {
+      retrieveSchemaFromApoc
+    } catch {
+      case e: ClientException =>
+        e.code match {
+          case "Neo.ClientError.Procedure.ProcedureNotFound" => retrieveSchema
+        }
+    })
+      .sortBy(t => t.name)
+
+    structFields += StructField(Neo4jQuery.INTERNAL_LABELS_FIELD, DataTypes.createArrayType(DataTypes.StringType), nullable = true)
+    structFields += StructField(Neo4jQuery.INTERNAL_ID_FIELD, DataTypes.IntegerType, nullable = false)
+    StructType(structFields.reverse)
+  }
+
+  private def retrieveSchemaFromApoc: mutable.Buffer[StructField] = {
+    session.run(
+      "CALL apoc.meta.nodeTypeProperties({ includeLabels: $labels })",
```

Comment:
PLACEHOLDER -- doesn't have to be handled in this PR.   This particular call along with relTypeProperties() has other parameters that it takes which are important.  They control sampling and metadata exclusion, among other things.    I think we may want a setting that allows a user to pass parameters here, to control how the metadata sampling is done.  On very large stores, this can get intensive without the right settings.

Reply:
Totally agree! We are aware of this, we mentioned it in the [estimate doc](https://docs.google.com/document/d/1qG0Wf8whkns3u6UAgrQTVaxGerEyCZHThNdmkVr3-UA/edit#). There is a lot going on in this PR so had to keep something out (like partitioning).